contrib/debian-template/rules

```diff
@@ -0,0 +1,38 @@
+#!/usr/bin/make -f
+
+include /usr/share/dpkg/default.mk
+
+export DEB_BUILD_MAINT_OPTIONS = hardening=+all optimize=+lto
+
+CUDA_RUNTIME_HEADER_PROVIDER = $(shell dpkg-query -S "*cuda_runtime.h" | cut -d':' -f1 | head -n1 | xargs)
+CUDA_LIB_PROVIDER = $(shell dpkg-query -S "libcudart.so.12" | cut -d':' -f1 | head -n1 | xargs)
+CUDA_SHLIB = $(shell dpkg-query -f '$(CUDA_LIB_PROVIDER) (>= ${source:Version})' -W $(CUDA_RUNTIME_HEADER_PROVIDER))
+FABRIC_HEADER_PROVIDER = $(shell (dpkg-query -S "/opt/amazon/efa/include/rdma/fabric.h" 2>/dev/null || dpkg-query -S "rdma/fabric.h") | cut -d':' -f1 | head -n1 | xargs)
+FABRIC_SHLIB = $(shell dpkg-query -f 'libfabric1-aws (>= ${source:Version})' -W $(FABRIC_HEADER_PROVIDER))
+
+override_dh_gencontrol:
+	echo "shlib:Depends=$(CUDA_SHLIB)" >> debian/libnccl-net-ofi-aws.substvars
+	echo "shlib:Depends=$(FABRIC_SHLIB)" >> debian/libnccl-net-ofi-aws.substvars
+	dh_gencontrol
+
+override_dh_missing:
+	find debian/ -name *.la -delete
+	dh_missing --fail-missing
+
+override_dh_shlibdeps:
+	echo "libfabric 1 libfabric1-aws (>= 1.22.0amzn2.0)" >> debian/shlibs.local
+	echo "libcudart 12 $(CUDA_LIB_PROVIDER)" >> debian/shlibs.local
```

Comment:
While technically correct, we actually need to not have a dependency on libcudart's package.  If the runscript installation method is used, dpkg isn't going to know about the libcudart.so file.

Reply:
Why bother with debian native packaging if we're not going to actually declare our dependencies?

Reply:
There's no win here.  But this is what we're doing with Libfabric and what we need to do here.  Stop being so pedantic and look at the customer picture.  You can't require them to use the native packaging installation, when most use the run script today.

Reply:
because that's how the efa installer works.

Reply:
The way that debian native packaging works is that if a library links another library, every DT_NEEDED entry on the binary needs to be resolved from some package that ships a file like:

```
$ dpkg-query -c libhwloc15               
/var/lib/dpkg/info/libhwloc15:amd64.md5sums
/var/lib/dpkg/info/libhwloc15:amd64.shlibs
/var/lib/dpkg/info/libhwloc15:amd64.triggers
$ cat "/var/lib/dpkg/info/libhwloc15:amd64.shlibs"

libhwloc 15 libhwloc15 (>= 2.10.0)
```

One can disable this completely on specific libraries with the [dh_shlibdeps](https://man7.org/linux/man-pages/man1/dh_shlibdeps.1.html) `-X` option, but it means that we then need to manually maintain dependency information for libfabric and hwloc.

Reply:
I'd also note that libfabric-aws does not do this correctly and/or doesn't ship a symbols file.

Reply:
Likely this is due to installing under /opt:

```
The package should install the shared libraries under their normal names. For example, the libgdbm3 package should install libgdbm.so.3.0.0 as /usr/lib/libgdbm.so.3.0.0. The files should not be renamed or re-linked by any prerm or postrm scripts; dpkg will take care of renaming things safely without affecting running programs, and attempts to interfere with this are likely to lead to problems.
```

from https://www.debian.org/doc/debian-policy/ch-sharedlibs.html